base/internal/message.hpp

```diff
@@ -89,6 +92,19 @@ inline std::string DebugPrint(std::u32string_view utf32)
   return internal::ToUtf8(utf32);
 }
 
+inline std::string DebugPrint(char32_t const * t)
+{
+  if (t) {
+    return DebugPrint(std::u32string(t));
+  }
+  return {"NULL string pointer"};
+}
+
+inline std::string DebugPrint(char32_t t)
+{
+  return DebugPrint(std::u32string(1, t));
+}
```

Comment:
I don't have c++20 yet but suspect that we can remove it for now.

Reply:
They do serve a purpose, the codebase doesn't simply compile when removed:

````
FAILED: base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx.o 
/usr/bin/c++ -DDEBUG -I/tmp/organicmaps -I/tmp/organicmaps/build_cpp20/3party/gflags/include -I/tmp/organicmaps/3party/fast_double_parser/include -isystem /tmp/organicmaps/3party/boost -isystem /tmp/organicmaps/3party/utf8cpp/include -g1 -g -std=c++20 -fPIC   -fno-omit-frame-pointer -fdiagnostics-color=always -ffast-math -Wall -Wextra -Wpedantic -Wno-unused-parameter -MD -MT base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx.o -MF base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx.o.d -o base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx.o -c /tmp/organicmaps/build_cpp20/base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx
In file included from /tmp/organicmaps/base/assert.hpp:4,
                 from /tmp/organicmaps/base/base.cpp:3,
                 from /tmp/organicmaps/build_cpp20/base/CMakeFiles/base.dir/Unity/unity_0_cxx.cxx:4:
/tmp/organicmaps/base/internal/message.hpp: In instantiation of ‘std::string DebugPrint(const T&) [with T = char32_t; std::string = std::__cxx11::basic_string<char>]’:
/tmp/organicmaps/base/internal/message.hpp:236:20:   required from ‘std::string base::Message(const T&) [with T = char32_t; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/internal/message.hpp:243:39:   required from ‘std::string base::Message(const T&, const Args& ...) [with T = char [16]; Args = {char32_t}; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/lower_case.cpp:172:14:   required from here
/tmp/organicmaps/base/internal/message.hpp:55:7: error: use of deleted function ‘std::basic_ostream<char, _Traits>& std::operator<<(basic_ostream<char, _Traits>&, char32_t) [with _Traits = char_traits<char>]’
   55 |   out << t;
      |   ~~~~^~~~
In file included from /usr/include/c++/13/istream:41,
                 from /usr/include/c++/13/sstream:40,
                 from /usr/include/c++/13/chrono:45,
                 from /tmp/organicmaps/base/internal/message.hpp:5:
/usr/include/c++/13/ostream:609:5: note: declared here
  609 |     operator<<(basic_ostream<char, _Traits>&, char32_t) = delete;
      |     ^~~~~~~~
[11/224] Building CXX object routing_common/CMakeFiles/routing_common.dir/Unity/unity_0_cxx.cxx.o
In file included from /tmp/organicmaps/indexer/classificator.hpp:3,
                 from /tmp/organicmaps/routing_common/bicycle_model.cpp:3,
                 from /tmp/organicmaps/build_cpp20/routing_common/CMakeFiles/routing_common.dir/Unity/unity_0_cxx.cxx:4:
/tmp/organicmaps/indexer/drawing_rule_def.hpp:48:80: warning: arithmetic between enumeration type ‘feature::Layer’ and floating-point type ‘const double’ is deprecated [-Wdeprecated-enum-float-conversion]
   48 |                         kMinLayeredDepthFg = kBaseDepthFg + feature::LAYER_LOW * kLayerPriorityRange,
      |                                                             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/tmp/organicmaps/indexer/drawing_rule_def.hpp:58:86: warning: arithmetic between enumeration type ‘feature::Layer’ and floating-point type ‘const double’ is deprecated [-Wdeprecated-enum-float-conversion]
   58 |                         kMinLayeredDepthBg = kBaseDepthBgBySize + feature::LAYER_LOW * kLayerPriorityRange;
      |                                                                   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
[12/224] Building CXX object base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx.o
FAILED: base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx.o 
/usr/bin/c++ -DDEBUG -DOMIM_UNIT_TEST_DISABLE_PLATFORM_INIT -I/tmp/organicmaps -I/tmp/organicmaps/build_cpp20/3party/gflags/include -isystem /tmp/organicmaps/3party/boost -isystem /tmp/organicmaps/3party/utf8cpp/include -g1 -g -std=c++20 -fPIE   -fno-omit-frame-pointer -fdiagnostics-color=always -ffast-math -Wall -Wextra -Wpedantic -Wno-unused-parameter -MD -MT base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx.o -MF base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx.o.d -o base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx.o -c /tmp/organicmaps/build_cpp20/base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx
In file included from /tmp/organicmaps/base/exception.hpp:3,
                 from /tmp/organicmaps/testing/testing.hpp:5,
                 from /tmp/organicmaps/base/base_tests/assert_test.cpp:1,
                 from /tmp/organicmaps/build_cpp20/base/base_tests/CMakeFiles/base_tests.dir/Unity/unity_0_cxx.cxx:4:
/tmp/organicmaps/base/internal/message.hpp: In instantiation of ‘std::string DebugPrint(const T&) [with T = char32_t; std::string = std::__cxx11::basic_string<char>]’:
/tmp/organicmaps/base/internal/message.hpp:243:20:   required from ‘std::string base::Message(const T&, const Args& ...) [with T = char32_t; Args = {char32_t}; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/base_tests/string_utils_test.cpp:81:7:   required from here
/tmp/organicmaps/base/internal/message.hpp:55:7: error: use of deleted function ‘std::basic_ostream<char, _Traits>& std::operator<<(basic_ostream<char, _Traits>&, char32_t) [with _Traits = char_traits<char>]’
   55 |   out << t;
      |   ~~~~^~~~
In file included from /usr/include/c++/13/istream:41,
                 from /usr/include/c++/13/sstream:40,
                 from /usr/include/c++/13/chrono:45,
                 from /tmp/organicmaps/base/internal/message.hpp:5:
/usr/include/c++/13/ostream:609:5: note: declared here
  609 |     operator<<(basic_ostream<char, _Traits>&, char32_t) = delete;
      |     ^~~~~~~~
/tmp/organicmaps/base/internal/message.hpp: In instantiation of ‘std::string DebugPrint(const T&) [with T = const char32_t*; std::string = std::__cxx11::basic_string<char>]’:
/tmp/organicmaps/base/internal/message.hpp:243:20:   required from ‘std::string base::Message(const T&, const Args& ...) [with T = const char32_t*; Args = {char32_t*}; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/buffer_vector.hpp:366:7:   required from ‘void buffer_vector<T, N>::insert(const_iterator, TIt, TIt) [with TIt = const char32_t*; T = char32_t; long unsigned int N = 32; const_iterator = const char32_t*]’
/tmp/organicmaps/base/string_utils.hpp:74:13:   required from ‘void strings::UniString::Replace(buffer_vector<char32_t, 32>::iterator, buffer_vector<char32_t, 32>::iterator, Iter, Iter) [with Iter = const char32_t*; buffer_vector<char32_t, 32>::iterator = char32_t*]’
/tmp/organicmaps/base/base_tests/string_utils_test.cpp:1227:22:   required from here
/tmp/organicmaps/base/internal/message.hpp:55:7: error: use of deleted function ‘std::basic_ostream<char, _Traits>& std::operator<<(basic_ostream<char, _Traits>&, const char32_t*) [with _Traits = char_traits<char>]’
   55 |   out << t;
      |   ~~~~^~~~
/usr/include/c++/13/ostream:703:5: note: declared here
  703 |     operator<<(basic_ostream<char, _Traits>&, const char32_t*) = delete;
      |     ^~~~~~~~
/tmp/organicmaps/base/internal/message.hpp: In instantiation of ‘std::string DebugPrint(const T&) [with T = char32_t*; std::string = std::__cxx11::basic_string<char>]’:
/tmp/organicmaps/base/internal/message.hpp:236:20:   required from ‘std::string base::Message(const T&) [with T = char32_t*; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/internal/message.hpp:243:39:   required from ‘std::string base::Message(const T&, const Args& ...) [with T = const char32_t*; Args = {char32_t*}; std::string = std::__cxx11::basic_string<char>]’
/tmp/organicmaps/base/buffer_vector.hpp:366:7:   required from ‘void buffer_vector<T, N>::insert(const_iterator, TIt, TIt) [with TIt = const char32_t*; T = char32_t; long unsigned int N = 32; const_iterator = const char32_t*]’
/tmp/organicmaps/base/string_utils.hpp:74:13:   required from ‘void strings::UniString::Replace(buffer_vector<char32_t, 32>::iterator, buffer_vector<char32_t, 32>::iterator, Iter, Iter) [with Iter = const char32_t*; buffer_vector<char32_t, 32>::iterator = char32_t*]’
/tmp/organicmaps/base/base_tests/string_utils_test.cpp:1227:22:   required from here
/tmp/organicmaps/base/internal/message.hpp:55:7: error: use of deleted function ‘std::basic_ostream<char, _Traits>& std::operator<<(basic_ostream<char, _Traits>&, const char32_t*) [with _Traits = char_traits<char>]’
   55 |   out << t;
      |   ~~~~^~~~
/usr/include/c++/13/ostream:703:5: note: declared here
  703 |     operator<<(basic_ostream<char, _Traits>&, const char32_t*) = delete;
      |     ^~~~~~~~
[16/224] Building CXX object geometry/geometry_tests/CMakeFiles/geometry_tests.dir/Unity/unity_0_cxx.cxx.o
ninja: build stopped: subcommand failed.
`````

Reply:
I see, ok keep only ```inline std::string DebugPrint(char32_t t)``` ?

Reply:
Should this function be removed?